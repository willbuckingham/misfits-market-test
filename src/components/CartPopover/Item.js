import React from 'react'
import PropTypes from 'prop-types'
import {
  Typography,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'

import { formatCurrency } from './../../utils'

const useStyles = makeStyles(({ spacing }) => ({
  root: {
    display: 'flex',
  },
  textarea: {
    padding: spacing(1),
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    flexGrow: 1,
  },
  price: {
    display: 'flex',
    alignItems: 'center',
  },
}))

const Item = props => {
  const classes = useStyles()

  const { product, price, quantity } = props

  return (
    <>
      <div className={classes.root}>
        <img src='/fruit.jpg' alt={product} title={product} height={80} />
        <div className={classes.textarea}>
          <Typography>{product}</Typography>
          <Typography color='textSecondary'>{formatCurrency(price)} x {quantity}</Typography>
        </div>
        <div className={classes.price}>
          <Typography variant='h6'>{formatCurrency(price * quantity)}</Typography>
        </div>
      </div>
      <hr/>
    </>
  )
}

Item.propTypes = {
  product: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
}

export default Item
