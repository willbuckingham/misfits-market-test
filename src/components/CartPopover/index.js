import CardPopover from './CartPopover'

export default CardPopover
