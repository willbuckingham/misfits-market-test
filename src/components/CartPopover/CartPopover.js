import React, { useState } from 'react'
import propTypes from 'prop-types'
import {
  Badge,
  IconButton,
  Popover,
  Button,
  Typography,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { 
  ShoppingCart,
} from '@material-ui/icons'

import Item from './Item'
import { formatCurrency } from '../../utils'

const useStyles = makeStyles(({ spacing }) => ({
  root: {
    maxWidth: '100%',
  },
  content: {
    padding: spacing(1),
    width: 500,
    maxWidth: '100%',
  },
  empty: {
    margin: spacing(8),
  },
}))

const CartPopover = ({ cartItems }) => {
  const classes = useStyles()

  const [anchorEl, setAnchorEl] = useState(null)
  const open = Boolean(anchorEl)

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  const isEmpty = !cartItems.length
  const count = cartItems.reduce((a,v) => a + parseInt(v.quantity), 0) || 0
  const total = cartItems.reduce((a,v) => a + v.quantity * v.price, 0) || 0

  return (
    <>
      <IconButton color='inherit' onClick={ handleClick }>
        <Badge badgeContent={count} color='secondary'>
          <ShoppingCart/>
        </Badge>
      </IconButton>

      <Popover 
        open={open}
        onClose={handleClose}
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'right',
        }}
      >
        <div className={classes.content}>
          <Typography variant='h6'>Items ({count})</Typography>
          <hr/>
          { isEmpty ?
            <Typography variant='body1' align='center' className={classes.empty}>Cart Empty!</Typography> :
            cartItems.map(v => <Item key={v.id} { ...v } />)
          }
          <Button
            variant='contained'
            color='primary'
            disabled={isEmpty}
            fullWidth
            onClick={() => alert('not broken, done in next Ticket ;-)')}
          >
            <Typography>Checkout { !!total && formatCurrency(total) }</Typography>
          </Button>
        </div>
      </Popover>
    </>
  )
}

CartPopover.propTypes = {
  cartItems: propTypes.array,
}

CartPopover.defaultProps = {
  cartItems: [],
}

export default CartPopover
