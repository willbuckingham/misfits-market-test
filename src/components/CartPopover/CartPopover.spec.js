import React from 'react'
import { shallow } from 'enzyme'

import CartPopover from './CartPopover'
import {
  IconButton,
  Popover,
} from '@material-ui/core'

describe('CartPopover', () => {
  let props

  beforeAll(() => {
    props = {
      cartItems: [],
    }
  })

  it('renders empty', () => {
    expect(shallow(<CartPopover { ...props } />)).toMatchSnapshot()
  })

  it('renders w/ item', () => {
    expect(shallow(<CartPopover { ...props } cartItems={[{ id: 1, product: 'p', price: 1, quantity: 2 }]} />)).toMatchSnapshot()
  })

  it('opens/closes popover', () => {
    const wrapper = shallow(<CartPopover { ...props } />)

    wrapper.find(IconButton).props().onClick({ currentTarget: 'mocktarget' })

    expect(wrapper.find(Popover).props().open).toBeTruthy()

    wrapper.find(Popover).props().onClose()

    expect(wrapper.find(Popover).props().open).toBeFalsy()
  })
})
