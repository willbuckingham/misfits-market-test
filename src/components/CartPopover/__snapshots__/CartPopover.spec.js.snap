// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPopover renders empty 1`] = `
<Fragment>
  <WithStyles(ForwardRef(IconButton))
    color="inherit"
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(Badge))
      badgeContent={0}
      color="secondary"
    >
      <ShoppingCartIcon />
    </WithStyles(ForwardRef(Badge))>
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(Popover))
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <div
      className="makeStyles-content-2"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        Items (
        0
        )
      </WithStyles(ForwardRef(Typography))>
      <hr />
      <WithStyles(ForwardRef(Typography))
        align="center"
        className="makeStyles-empty-3"
        variant="body1"
      >
        Cart Empty!
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={true}
        fullWidth={true}
        onClick={[Function]}
        variant="contained"
      >
        <WithStyles(ForwardRef(Typography))>
          Checkout 
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Button))>
    </div>
  </WithStyles(ForwardRef(Popover))>
</Fragment>
`;

exports[`CartPopover renders w/ item 1`] = `
<Fragment>
  <WithStyles(ForwardRef(IconButton))
    color="inherit"
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(Badge))
      badgeContent={2}
      color="secondary"
    >
      <ShoppingCartIcon />
    </WithStyles(ForwardRef(Badge))>
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(Popover))
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <div
      className="makeStyles-content-2"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        Items (
        2
        )
      </WithStyles(ForwardRef(Typography))>
      <hr />
      <Item
        id={1}
        key="1"
        price={1}
        product="p"
        quantity={2}
      />
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={false}
        fullWidth={true}
        onClick={[Function]}
        variant="contained"
      >
        <WithStyles(ForwardRef(Typography))>
          Checkout 
          $2.00
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Button))>
    </div>
  </WithStyles(ForwardRef(Popover))>
</Fragment>
`;
